l1/l1.go

```diff
@@ -75,101 +62,86 @@ func (c *Client) checkChainID(ctx context.Context) error {
 	// Starknet to create a "custom" Starknet network, we will need to log a warning instead.
-	return fmt.Errorf("mismatched L1 and L2 networks: L2 network %s; is the L1 node on the correct network?", c.network)
+	return fmt.Errorf("mismatched L1 and L2 networks:  L1 chain ID: %s, L2 chain ID: %s,"+
+		" is the L1 node on the correct network (see --eth-node)?", wantChainID.Text(felt.Base16), gotChainID.Text(felt.Base16))
 }
 
 func (c *Client) Run(ctx context.Context) error {
-	defer c.l1.Close()
 	if err := c.checkChainID(ctx); err != nil {
 		return err
 	}
 
-	buffer := 128
-
-	logStateUpdateChan := make(chan *contract.StarknetLogStateUpdate, buffer)
-	subUpdates, err := c.subscribeToUpdates(ctx, logStateUpdateChan)
-	if err != nil {
-		return err
-	}
-	defer subUpdates.Unsubscribe()
-
-	headerChan := make(chan *types.Header, buffer)
-	subHeaders, err := c.subscribeToHeaders(ctx, headerChan)
-	if err != nil {
-		return err
-	}
-	defer subHeaders.Unsubscribe()
-
+	ticker := time.NewTicker(c.delay)
+	defer ticker.Stop()
 	for {
 		select {
 		case <-ctx.Done():
 			return nil
-		case err := <-subHeaders.Err():
-			c.log.Warnw("L1 header subscription failed, resubscribing", "error", err)
-			subHeaders.Unsubscribe()
+		case <-ticker.C:
+			latestL1BlockNumber, finalisedL1BlockNumber, latestL1BlockHash, err := c.l1.LatestAndFinalisedHeights(ctx)
+			if err != nil {
+				c.log.Warnw("Failed to fetch finalised and latest L1 block numbers", "err", err)
+				continue
+			}
 
-			subHeaders, err = c.subscribeToHeaders(ctx, headerChan)
+			if c.finalisedHeight > finalisedL1BlockNumber {
+				// TODO how should we handle such catastrophic events?
+				return errReorgFinalisedBlock
+			}
+
+			c.finalisedHeight = finalisedL1BlockNumber
+
+			head, err := c.l1.StarknetHead(ctx, latestL1BlockHash)
 			if err != nil {
-				return err
+				c.log.Warnw("Failed to fetch Starknet L1 head", "err", err)
+				continue
 			}
-			defer subHeaders.Unsubscribe() //nolint:gocritic
-		case header := <-headerChan:
-			l1Height := header.Number.Uint64()
-			c.log.Debugw("Received L1 header", "number", l1Height, "hash", header.Hash().Hex())
-		Outer:
-			// Check for updates.
-			// We need to loop in case there were multiple LogStateUpdates emitted.
-			for {
-				select {
-				case err := <-subUpdates.Err():
-					c.log.Warnw("L1 update subscription failed, resubscribing", "error", err)
-					subUpdates.Unsubscribe()
-
-					subUpdates, err = c.subscribeToUpdates(ctx, logStateUpdateChan)
-					if err != nil {
-						return err
+			c.log.Debugw("Retrieved L1 head", "latestL1", latestL1BlockNumber, "finalisedL1",
+				finalisedL1BlockNumber, "l2BlockNumber", head.BlockNumber)
+
+			c.finalisedQueue[latestL1BlockNumber] = head
+			var maxFinalisedNumber uint64
+			for l1BlockNumber := range c.finalisedQueue {
+				if l1BlockNumber > latestL1BlockNumber {
+					// Ethereum block was reorged. Remove l1Head from queue.
+					delete(c.finalisedQueue, l1BlockNumber)
```

Comment:
What we should do is, store `latestL1BlockHash` together with the head here; 

`c.finalisedQueue[latestL1BlockNumber] = head`

and make sure that `latestL1BlockHash` is still a part of the canonical chain.

Reply:
Honestly, I believe the event subscription approach handles this better. Instead of `latest - confirmationPeriod` we would just check the events we have in the queue against the most recent `finalisedL1BlockNumber`

Reply:
Good point. I'll add tests for this.

To fix, I think I'll use [`eth_newBlockFilter`](https://docs.infura.io/networks/ethereum/json-rpc-methods/filter-methods/eth_newblockfilter) + [`eth_getFilterChanges`](https://docs.infura.io/networks/ethereum/json-rpc-methods/filter-methods/eth_getfilterchanges) to maintain a slice of block hashes from the finalized block to the latest block. i.e., we keep track of the fork the Ethereum node is on.

Reply:
You were right, the event subscriptions are the way to go. Updated.

Reply:
> In practice ethereum nodes never rollback. They just switch to a better fork with atleast the same amount of blocks they have locally, so latestL1BlockNumber will always monotonically increase.

As far as I understand the above is not true because reorg logic is not "longest chain" rather it is the "heaviest chain". Therefore, it is entirely possible for an ethereum to switch to a chain with a lower height but it is "heavier" than a "longer" chain. For `ethash`, the reorg logic would take previous blocks + uncle/omner blocks into account to decide the chain to build on. I assume this was kept for the DAG part of Casper consensus mechanism.